src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp

```diff
@@ -348,11 +358,13 @@ void LinalgBackendEigen::add_scalar_impl(SGMatrix<T>& a, T b) const
 	a_eig = a_eig.array() + b;
 }
 
-template <typename T>
-T LinalgBackendEigen::dot_impl(const SGVector<T>& a, const SGVector<T>& b) const
+template <typename T, typename U, typename TU>
+TU LinalgBackendEigen::dot_impl(const SGVector<T>& a, const SGVector<U>& b) const
 {
-	return (typename SGVector<T>::EigenVectorXtMap(a))
-	    .dot(typename SGVector<T>::EigenVectorXtMap(b));
+	typename SGVector<T>::EigenMatrixXtMap a_eig = a;
+	typename SGVector<U>::EigenMatrixXtMap b_eig = b;
+
+	return a_eig.template cast<TU>().dot(b_eig.template cast<TU>());
```

Comment:
Could you check whether eigen SIMDs this if the types are the same?
I know things are casted on the fly, but that makes the whole thing slower if done when the types are the same

Reply:
yeah I checked, same performance with O3 optimization. It gets much slower when using different types
```
2019-06-13 23:32:58
Running ./out
Run on (4 X 3000 MHz CPU s)
CPU Caches:
  L1 Data 32K (x2)
  L1 Instruction 32K (x2)
  L2 Unified 256K (x2)
  L3 Unified 4096K (x1)
Load Average: 2.77, 4.82, 5.70
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
BM_std_dot                  10237 ns        10233 ns        67238
BM_dot                       3654 ns         3654 ns       190775
BM_dot_without_casting       3644 ns         3644 ns       193444
BM_loop_dot                 30550 ns        30549 ns        22922
```
BM_dot is eigen without casting. std_dot is std::inner_product. This is the test case
```
static void BM_dot(benchmark::State& state)
{
	SGVector<float64_t> x(10000);
	x.range_fill(1);

	SGVector<float64_t> y(10000);
	y.set_const(1);

	for (auto _ : state)
	{
		benchmark::DoNotOptimize(dot(x, y));
	}
}
```

Reply:
Ok thanks!
I wonder: If we take such a performance penalty, then why offer this?

Reply:
Eigen is still faster than using for loops with different types. Also it will make things more elegant (as in DenseFeatures where loops are used when different types are multiplied).

Reply:
ok cool. Just saying though, people tend to use code without thinking about it. So sometimes it might be better to not allow certain things, and force people to be more explicit in their code. But here with the features, ok makes sense, just looking at the other Pr

Reply:
should we make it harder to use this? like by passing a "slow_dot" tag or something

Reply:
I think that might be an idea. Only question is whether this is overkill ;)
But let’s try

Reply:
@karlnapf It looks ugly. I have to ask about the type in the tag in LinalgNamespace, and then dispatch accordingly since the backend can't do that

Reply:
Maybe a decorator that one hs to pass to avoid compiler warning? Not sure that is even possible....

Reply:
why don't you have an allow_cast templated bool that is by default false and if you have to cast it is a compiler error and then a user has to set it as true if he wants to cast/knows he has to cast?

Reply:
@gf712  That is more elegant, but it's runtime. I found the compile time solution ugly since LinalgNamespace has to make the decision not the backend

Reply:
you mean you need something for runtime?